### ISSUE TITLE
Plugable Compression

### ISSUE THREAD

**Issue by veblush** [grpc/grpc#29767]
Currently gRPC C++ supports built-in compression algorithms. As more compelling algorithms to come, gRPC needs to accommodate those but it's not a good idea to add those to builtin one because it will increase footprints and dependency. Rather, we can devise a pluggable compression system so that users can register their own ones in advance. This is already available on gRPC Java and Go so we can do the same thing for gRPC Core / C++.

**Comment by elftech** [comment]
Awesome! Looking forward to this feature.

I have personally added some new compression algorithms to my local  grpc repository  successfully before.  But the developing process turned to be, unexpectedly, a bit of annoying due to  grpc's split of raw source data,  which may take more handling when the streaming compression interface is not friendly enough.

Moreover, would you add more default supported algorithm to this pluggable system?

**Comment by veblush** [comment]
@elftech gRPC doesn't need to add more builtin algorithm if there is a plugable compression system allowing to bring your own things without increasing the footprint of the gRPC library.

**Comment by bistack** [comment]
Is there any workable lz4 MR?

**Comment by psychobug420** [comment]
Is this still relevant, I would be happy to work on this?

If there any details I should consider let me know please

**Comment by pawbhard** [comment]
@psychobug420 Yes it is relevant, please start with the API proposal in (grpc/proposal) which can help in plugging different compressions, let us review that, before you start implementing it.

**Comment by psychobug420** [comment]
Thanks for this information, I will prepare the proposal, open a PR and post it here.

**Comment by psychobug420** [comment]
Hello, while working on proposal I got one question I don't have an answer for.

For context: I am new to gRPC and I see that gRPC has some languages supported as native, for example native implementations for Java, .NET, Go and some others. Also there is 2nd type of gRPCs, for example Python or PHP ones, which are bindings to C++ implementation via C wrapper, this is the one I am working on a proposal.

I am getting right that it should be supported cross language, by this I mean the languages which are bound to C++ implementation?

**Comment by pawbhard** [comment]
Yes. we need for C++ and wrapped languages.

**Comment by psychobug420** [comment]
Should we support multiple custom compression algorithms simultaneously?

I was playing around with the library, and thought about use cases where this would matter. Like, an app running multiple gRPC channels where each needs its own compression algorithm, or wanting specific message types compressed differently. Asking this because supporting different algorithms would affect implementation complexity a lot

P.S.
Sorry for the slow reply, been swamped a bit and couldn't get back to this earlier
